libkbfs/impatient_debug_dumper.go

```diff
@@ -0,0 +1,213 @@
+package libkbfs
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"runtime/pprof"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/keybase/client/go/logger"
+	"golang.org/x/time/rate"
+)
+
+type ctxTimeTracker struct {
+	ctx       context.Context
+	expiresAt time.Time
+
+	done int32
+}
+
+func (c *ctxTimeTracker) markDone() {
+	atomic.StoreInt32(&c.done, 1)
+}
+
+func (c *ctxTimeTracker) isDone() bool {
+	return atomic.LoadInt32(&c.done) == 1
+}
+
+func (c *ctxTimeTracker) isExpired(clock Clock) bool {
+	return clock.Now().After(c.expiresAt)
+}
+
+type ctxTimeTrackerListNode struct {
+	tracker *ctxTimeTracker
+	next    *ctxTimeTrackerListNode
+}
+
+type ctxTimeTrackerList struct {
+	front *ctxTimeTrackerListNode
+	tail  *ctxTimeTrackerListNode
+}
+
+func (c *ctxTimeTrackerList) peekFront() (t *ctxTimeTracker) {
+	if c.front == nil {
+		return nil
+	}
+	return c.front.tracker
+}
+
+func (c *ctxTimeTrackerList) popFront() (t *ctxTimeTracker) {
+	if c.front == nil {
+		return nil
+	}
+	t = c.front.tracker
+	c.front = c.front.next
+	if c.front == nil {
+		// last one!
+		c.tail = nil
+	}
+	return t
+}
+
+func (c *ctxTimeTrackerList) append(t *ctxTimeTracker) {
+	newNode := &ctxTimeTrackerListNode{
+		tracker: t,
+	}
+	if c.tail != nil {
+		c.tail.next = newNode
+	}
+	c.tail = newNode
+	if c.front == nil {
+		c.front = newNode
+	}
+}
+
+// ImpatientDebugDumper dumps all running goroutines if an operation takes
+// longer than a preset value. User of this type should call Begin() with a
+// context associated with an operation, and call the returned function when
+// the operation is done. If the operation finishes within the preset duration,
+// nothing is dumped into log. Despite being impatient, it tries not to pollute
+// the log too much by rate limit goroutine dumping based on
+// impatientDebugDumperDumpMinInterval (at most 1 per minute).
+type ImpatientDebugDumper struct {
+	config Config
+	log    logger.Logger
+	dumpIn time.Duration
+
+	ticker                 *time.Ticker
+	limiter                *rate.Limiter
+	idempotentShutdownFunc func()
```

Comment:
You can do it this way if you want, but naming it `idempotent` doesn't really add anything. I'd prefer either `shutdownFunc`/`shutdownFn` or `shutdownCh` for a channel closed in a `sync.Once`.